### DISCUSSION TITLE
MS SQL 2014 - Scatter of results

### DISCUSSION THREAD

**Discussion by everande** [TPC-Council/HammerDB#294] [Q&A] [answered]
Hello,

What could be the reason for the non-repeatability of the results?

I have a new server. There is only one VM to use with HammerDB, so HammerDB is the only one workload. I have made three tests - each time restore tpcc database, restart the SQL service to have empty RAM, and so on. I'm using 12 cores (threads). The results looks like below - each time results are quite different when more users are working on database:

![TPM](https://user-images.githubusercontent.com/96062550/145822999-a82bdb69-45eb-4848-8391-f38fd39854f2.JPG)

**Comment by sm-shaw** [comment, marked as answer]
There are a wide number of reasons to why you could see variability in the workload. First of course, you are looking at an average of the TPM over a measured period of time. So the first thing you should do is to check the throughput at the higher core count to see if the transaction count is relatively level or whether you see peaks and troughs indicating contention.  If you see contention such use a tool such as activity monitor and identify your top wait events. 

![sqltpm](https://user-images.githubusercontent.com/38044085/145858630-96ba65d1-ea1c-4a59-95b5-356e560be8f7.PNG)

**Reply by everande** [reply]
thanks for the answer

I don't understand an idea - for me because it is an average of the TMP over a measured period of time it should be more stable / more repetitive and not like this, when one of measure-steps is... unreal looking at the curve (the is no other databases, services, applications in given VM and no other VM's on the physical server; HammerDB is running on the server - not on the client)

![HammerDB2](https://user-images.githubusercontent.com/96062550/147588601-f7d4d7f5-82e7-4e89-b95c-eda9a9a89315.JPG)

**Reply by sm-shaw** [reply]
It is about understanding how a database works and identifying what you are waiting on that is important.  Look at the graphical transaction counter for a run, or use the CLI equivalent to ensure that the results are consistent. Use the Activity Monitor to identify the performance bottlenecks - for example, can the  transaction log can sustain the high levels of throughput e.g. https://www.hammerdb.com/blog/uncategorized/how-to-improve-redo-transaction-log-and-wal-throughput-for-hammerdb-benchmarks/.  Any database will run inconsistently if it is not configured correctly.

**Reply by everande** [reply]
Of course, I will think over what you wrote

However, I am surprised by the volatility of the results, e.g. below chart. Each of the two test sequences (1 2 4 8 ... users) started as it seems under the same conditions, i.e. tpcc DB restor, restart of SQL service, LOG file on separate RAID1 drive, etc. 

![HammerDB3](https://user-images.githubusercontent.com/96062550/147605977-31a79c39-1688-4593-b0df-cea7d63900b7.JPG)

Single "erroneous" results (deviations) as on previous chart is one issue, but another issue seems to be completely different characteristics of the curve in two different trials under the same conditions?

**Reply by everande** [reply]
OK, I can see something like this:

There is continuous LOG file write only. Periodically there is data write do ndf files which reduces the number of transactions. I'm wondering why in some trial this is more often than in other (same disk, same data file location, same begnining conditions)

![HammerDB5](https://user-images.githubusercontent.com/96062550/147608552-0d9d24d4-0c22-4d85-8aa8-5348c240c57e.jpg)

![HammerDB4](https://user-images.githubusercontent.com/96062550/147608559-e8565bcf-6b84-4f55-b8fa-68a916671157.jpg)

**Reply by sm-shaw** [reply]
Great - this is called a checkpoint
https://docs.microsoft.com/en-us/sql/relational-databases/logs/database-checkpoints-sql-server?view=sql-server-ver15
There is a tips/best practice document, please note that this quite old and not all of the advice is up to date but it gives guidance on checkpoints here:
https://www.hammerdb.com/blog/uncategorized/hammerdb-best-practice-for-sql-server-performance-and-scalability/

**Reply by everande** [reply]
Thanks for your patience ;)

The checkpoint isn't maybe so strange, rather why once have impact and next time not - only this :)

BTW
About SQL server / DB configuration for HammerDB... I'm using configuration and properties similar to my production enviroment. Is optimising DB for HammerDB are right? The is no about to have higher result - higher number but real production databese performance, is it? At the end production DB can have different configuration forced by application spec or production app best practices

**Reply by sm-shaw** [reply]
The checkpoint is not unusual at all, when you are running multiple workloads, follow the guidance above and set DBCC trace 3502, 3504 and 3605 to see them occurring in the error log. 

For optimized vs production performance, this is one of the most common beginning errors. Optimized performance is generic, whereas production performance is very precise and only relevant to one scenario, preventing direct comparison between systems.  For one example, synchronous commit is always used in production, however using asynchronous commit will permit higher CPU utilization on a lower performing transaction log disk drive to simulate an environment where you have spent more on I/O capabilities. Significant investment in a production environment will give results similar to the optimized one but cost you less. The optimized results 'sets the bar' for which performance should never be exceeded in the production environment.  

I would recommend start with an optmized configuration then test the production settings to measure the impact of the production configuration and find where you could invest more for better performance in CPU, memory and I/O.

**Reply by everande** [reply]
thanks for all your replies - very helpful
